File: clients/drive9-js/examples/vault.ts
import { Client } from "../src/index.js";

async function main() {
  const client = Client.defaultClient();
  // Or: const client = new Client("http://127.0.0.1:9009", "your-api-key");

  const secretName = "example-secret";

  // Create or update
  await client.createVaultSecret(secretName, {
    password: "hunter2",
    api_key: "d9_xxxxxxxx",
  });
  console.log("Created vault secret:", secretName);

  // Read
  const secret = await client.readVaultSecret(secretName);
  console.log("Read secret:", secret);

  // Read single field
  const field = await client.readVaultSecretField(secretName, "password");
  console.log("Password field:", field);

  // List
  const secrets = await client.listVaultSecrets();
  console.log("Vault secrets:");
  for (const s of secrets) {
    console.log(`  ${s.name}`);
  }

  // Issue grant (spec §6)
  const grant = await client.issueVaultToken("agent-1", [secretName], "read", 3600);
  console.log("Issued grant:", grant.grant_id, "perm:", grant.perm, "ttl:", grant.ttl);

  // Revoke
  await client.revokeVaultToken(grant.grant_id);
  console.log("Revoked grant");

  // Clean up
  await client.deleteVaultSecret(secretName);
  console.log("Deleted secret");
}

main().catch(console.error);

File: clients/drive9-rs/examples/vault.rs
use drive9::Client;
use serde_json::json;

#[tokio::main]
async fn main() -> Result<(), drive9::Drive9Error> {
    let client = Client::default_client();
    // Or: let client = Client::new("http://127.0.0.1:9009", "your-api-key");

    let secret_name = "example-secret";

    // Create or update a secret
    let mut fields = serde_json::Map::new();
    fields.insert("password".to_string(), json!("hunter2"));
    fields.insert("api_key".to_string(), json!("d9_xxxxxxxx"));

    client.create_vault_secret(secret_name, &fields).await?;
    println!("Created vault secret: {}", secret_name);

    // Read the secret
    let secret = client.read_vault_secret(secret_name).await?;
    println!("Read secret: {:?}", secret);

    // Read a single field
    if let Ok(field) = client.read_vault_secret_field(secret_name, "password").await {
        println!("Password field: {}", field);
    }

    // List secrets
    let secrets = client.list_vault_secrets().await?;
    println!("Vault secrets:");
    for s in secrets {
        println!("  {}", s.name);
    }

    // Issue a capability grant (spec §6)
    let grant = client
        .issue_vault_token(
            "agent-1",
            &["example-secret".to_string()],
            "read",
            3600,
            Some("example-grant"),
        )
        .await?;
    println!("Issued grant: {} (perm={}, ttl={}s)", grant.grant_id, grant.perm, grant.ttl);

    // Revoke the grant
    client.revoke_vault_token(&grant.grant_id).await?;
    println!("Revoked grant");

    // Clean up
    client.delete_vault_secret(secret_name).await?;
    println!("Deleted secret");

    Ok(())
}
